### DISCUSSION TITLE
Quarto Jupyter: Unable to Import Modules

### DISCUSSION THREAD

**Discussion by kmcleste** [quarto-dev/quarto-cli#592] [Q&A] [answered]
I am currently trying to render a quarto markdown file that contains python code, which I would like to display using the Jupyter feature. I am also using this in conjunction with a Hugo site.

The problem I am facing is that every time I attempt to render the page, I get a ModuleNotFoundError as shown by the following traceback:

```bash
➜  workflow git:(main) ✗ source plaid-env/bin/activate
(plaid-env) ➜  workflow git:(main) ✗ quarto render plaid-ml.qmd   

Starting Jupyter kernel...Done

Executing 'plaid-ml.ipynb'
  Cell 1/2...ERROR: 

An error occurred while executing the following cell:
------------------
import numpy as np
from os import environ
environ["KERAS_BACKEND"] = "plaidml.keras.backend"
import keras
from keras.layers import Dense
import matplotlib.pyplot as plt
------------------

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Input In [1], in <module>
      2 from os import environ
      3 environ["KERAS_BACKEND"] = "plaidml.keras.backend"
----> 4 import keras
      5 from keras.layers import Dense
      6 import matplotlib.pyplot as plt

ModuleNotFoundError: No module named 'keras'
ModuleNotFoundError: No module named 'keras'
```

I have already verified that keras is installed:

```bash
(plaid-env) ➜  workflow git:(main) ✗ pip list | grep keras
keras               2.8.0
```

I have also successfully used the keras module in a jupyter notebook, with the interpreter set to use the same virtual environment that I have sourced.

Occasionally, I am also prompted to activate my virtual environment after it has already been sourced -- leading me to believe quarto may not be taking advantage of the environment and just using the default. 

I have also tried rendering just my jupyter notebook using Quarto, but am only able to output to HTML. Is it possible to configure this to convert to markdown?

Any suggestions would be much appreciated. Thanks

Quarto Version == 0.9.215

**Comment by jjallaire** [comment, marked as answer]
You appear to be doing things correctly (activating the virtual environment in the shell and then calling quarto render).

If you execute `quarto check jupyter` you'll see exactly which Python version/environment Quarto is binding against. This should be discovered from the `PATH` so should indeed work as you are expecting it to (note that Quarto only works w/ Python 3 so is explicitly looking for that not Python 2). You can force the version of Python used by Quarto via the `QUARTO_PYTHON` environment variable.

Note that one thing that *could* be happening is that in the loading of the Keras library its failing to bind to plaidml somehow, and that manifests as a "ModuleNotFound" (that's complete speculation on my part though).

To convert to markdown just do `quarto render notebook.ipynb --to markdown`. That will give you Pandoc markdown, you can also try `quarto render notebook.ipynb --to gfm` for Github Flavored Markdown.

**Reply by kmcleste** [reply]
Interesting - I ran `quarto check jupyter` and it shows:

```bash
[✓] Checking Python 3 installation....OK
      Version: 3.8.5
      Path: /Users/kmclester/.envs/plaid-env/bin/python3
      Jupyter: 4.9.2
      Kernels: python3

[✓] Checking Jupyter engine render....OK
```

This is the correct virtual environment but still no luck. I am also able to run the code chunks interactively using the Quarto VS Code extension. Additionally, I tried uninstalling the plaidml version of keras and then installed standalone keras version 2.8.0 and that did not work either.

Converting the notebook to markdown **does** work and I will likely use that method for the time being -- however, it would be nice to be able to render the qmd.

**Reply by jjallaire** [reply]
Wow very strange! I don't have further speculation on what could be going wrong w/ importing Keras.

**Reply by kmcleste** [reply]
So sorry for such a late reply. I was able to get it working by creating the virtual environment from within the VS Code terminal so that it automatically sourced the environment. Not quite sure **why** that works, but it does. I will mark this discussion as answered.

**Reply by jjallaire** [reply]
okay, great to hear!

**Comment by teuffy** [comment]
# Issue related to conda and env
## Setup
-  quarto --version 99.9.9

### conda list:
  - base
  - env

### jupyter kernelspec list
  - base : python397jvsc74a57bd0de1a830acc13051915ae3357edac6a95e4fd400b5fd3b09883bbcbf9b6b3fd74
  - env : python3

# Problem:

I notice when I activate env and install a library in, quarto preview can't see the library, but as it seems to look at the base environment by default (the one seen is 
 python397jvsc74a57bd0de1a830acc13051915ae3357edac6a95e4fd400b5fd3b09883bbcbf9b6b3fd74).

When I install the appropriate library in the base, quarto preview see the python library.

# Question:

- Is there an issue with quarto and conda with multiple env? 
- My way to solve it was to installed the libraries in the first seen by quarto when running preview (python397jvsc74a57bd0de1a830acc13051915ae3357edac6a95e4fd400b5fd3b09883bbcbf9b6b3fd74 in my case).
